src/Common/src/TypeSystem/Common/Utilities/GCPointerMap.cs

```diff
@@ -0,0 +1,265 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.TypeSystem
+{
+    /// <summary>
+    /// Represents a bitmap of GC pointers within a memory region divided into
+    /// pointer-sized cells.
+    /// </summary>
+    public struct GCPointerMap : IEquatable<GCPointerMap>
+    {
+        // Each bit in this array represents a pointer-sized cell.
+        private int[] _gcFlags;
+
+        private int _numCells;
+
+        /// <summary>
+        /// Gets a value indicating whether this map is uninitialized.
+        /// </summary>
+        public bool IsUninitialized
+        {
+            get
+            {
+                return _gcFlags == null;
+            }
+        }
+
+        /// <summary>
+        /// Gets the size (in cells) of the pointer map.
+        /// </summary>
+        public int Size
+        {
+            get
+            {
+                return _numCells;
+            }
+        }
+
+        public GCPointerMap(int[] gcFlags, int numCells)
+        {
+            Debug.Assert(numCells < gcFlags.Length << 5);
+            _gcFlags = gcFlags;
+            _numCells = numCells;
+        }
+
+        public static GCPointerMap FromInstanceLayout(DefType type)
```

Comment:
Cool, I pulled out the algorithm part in a side file so that it doesn't get in the way for consumers who don't want it. The type system algorithm will not try to cache then!